### DISCUSSION TITLE
Wrap shell output in code blocks

### DISCUSSION THREAD

**Discussion by talios** [Taitava/obsidian-shellcommands#244] [Ideas - done & released]
Often I find myself wanting to wrap the output of a shell command in a code-fence/preformatted code block for easier reading.  And maybe prefixing it with some custom templated output ( i.e run date etc. )

Is this something we can have added?

**Comment by Taitava** [comment]
Thank you for the suggestion! 🙂 Three things come to my mind:

## 1. Predefined ~~templates~~ wrappers
(Edit 2022-08-29: I'm trying to avoid using the word _template_ as that's usually used for markdown template files in Obsidian. _Wrapper_ is better, as was used in the original message by talios).

I think this is what you meant. Make it possible to configure text snippets that could be used as wrappers for output.
E.g. the following could be an output wrapper text:
````
```
{{output}}
```
````

If the executed shell command is e.g. `ls` (lists files and directories), then an example resulting output could be:
````
```
MyFolder MyNote.md MyOtherNote.md
```
````

The wrapper texts could be defined in settings, and multiple shell commands could use the same wrapper if wanted - no need to redefine the same wrapper for every shell command. The wrapper could be used regardless of what output channel is used: when outputting into file, into clipboard or into a notification balloon or whatever. Was this the thing you were asking for? I like it. 😎 

> And maybe prefixing it with some custom templated output ( i.e run date etc. )

I'm not sure if I understood correctly, did you mean something like this:

Use other variables in the wrapper text, e.g.:
````
Shell command result, ran on {{date:YYYY-MM-DD}}:
```
{{output}}
```
````


## 2. Reusable output blocks
Some kind of syntax to mark in a note file that _put output text here, and if the shell command is executed again, replace old output with new output (or add below the old output)_. This idea is still very new and unclear on my head, but something I'd like to think about and come up with a solution.

Main benefits would be to be able to define a permanent location for output in a note, and the ability to replace old output text. Then the location is not dependent on caret location; nor needed to be always outputted to top/bottom of the file.

## 3. Workaround for now: Do the wrapping in the shell command itself

Say that you have the following shell command that just lists all files and folders in the root of the Obsidian vault: `ls`
You can add two `echo` commands:

For macOS/Linux Bash:
```
echo \`\`\`
ls
echo \`\`\`
```

For Windows PowerShell:

```
echo ``````
ls
echo ``````
```

A downside here is that it makes the actual shell commands larger an harder to read, and you need to redefine the wrappers in every shell commands where you want to use them.

**Reply by Taitava** [reply]
# It's in progress
I made an issue about _1. Predefined wrappers_: #262

**Comment by Taitava** [comment]
# Output can be duplicated
Just a small and maybe not so important side effect that will come from this wrapper feature, is that if you use `{{output}}` variable many times, you can duplicate the shell command's output as many times as you wish (without executing the shell command many times).

An example wrapper:
````
DUPLICATED shell command result, ran on {{date:YYYY-MM-DD}}:
```
{{output}}
{{output}}
```
````

I don't know if there's any real use for this, though.

**Comment by Taitava** [comment]
# Released now
@talios Can you test this and let me know if there's anything to improve, please? 🙂

**Comment by talios** [comment]
@Taitava so far, after a quick test it seems to work wonderfully.

Only thing I can think of off hand, which could be handled at the level of the command level I'm running, but may be useful - is being able to trim the leading/trailing blank lines from any shell output.

That's a super minor/trivial thing tho.

**Reply by Taitava** [reply]
Thank you for your feedback! Yea, trimming blank lines from start/end of output has crossed my mind sometimes. It could be a logical extension for this feature. Some configuration parameters for users to decide:
- Trim from start/end only, or both.
- What to trim:
  - Newlines (`\r`, `\n` and `\r\n`)
  - Tabs
  - Spaces
  - A custom set of characters.

I don't know when I'm gonna implement this. My time is currently limited as I've started in a new job as a programmer a month ago, and there's a lot for me to learn in that project.

**Comment by talios** [comment]
> My time is currently limited as I've started in a new job as a programmer a month ago, and there's a lot for me to learn in that project.

Take your time! I know the feeling after doing deep-dives on pipelines and containers on my own projects at work.
